### PR TITLE
Add CPU support for torch and tensorflow dataloaders

### PR DESCRIPTION
**Pull request by jperez999**
this PR will setup CPU support for dataloaders

### CODE REVIEW

File: tests/unit/test_torch_dataloader.py [hunk 1]
```diff
@@ -245,7 +243,8 @@ def test_gpu_dl(tmpdir, df, dataset, batch_size, part_mem_fraction, engine, devi
         os.path.join(output_train, x) for x in os.listdir(output_train) if x.endswith("parquet")
     ]
 
-    nvt_data = nvt.Dataset(tar_paths[0], engine="parquet", part_mem_fraction=part_mem_fraction)
+    cpu_true = True if device =='cpu' else False
+    nvt_data = nvt.Dataset(tar_paths[0], cpu=cpu_true, engine="parquet", part_mem_fraction=part_mem_fraction)
```
Comment (benfred) [inline tests/unit/test_torch_dataloader.py:247]:
```suggestion
    cpu = device =="cpu"
    nvt_data = nvt.Dataset(tar_paths[0], cpu=cpu, engine="parquet", part_mem_fraction=part_mem_fraction)
```

File: nvtabular/loader/backend.py [hunk 2]
```diff
@@ -152,8 +157,10 @@ def start(self):
     def get_batch_div_chunk(self, chunks, batch_size):
         # TODO: is there a way to do this using cupy?
         spill_idx = int(chunks.shape[0] / batch_size) * batch_size
-        spill = cudf.DataFrame(chunks.iloc[spill_idx:])
-        chunks = cudf.DataFrame(chunks.iloc[:spill_idx])
+        #spill = cudf.DataFrame(chunks.iloc[spill_idx:])
+        spill = _make_df(chunks.iloc[spill_idx:])
+        #chunks = cudf.DataFrame(chunks.iloc[:spill_idx])
+        chunks = _make_df(chunks.iloc[:spill_idx])
```
Comment (benfred) [inline nvtabular/loader/backend.py:163]:
```suggestion
        spill = _make_df(chunks.iloc[spill_idx:])
        chunks = _make_df(chunks.iloc[:spill_idx])
```

File: nvtabular/loader/backend.py [hunk 3]
```diff
@@ -21,11 +21,11 @@
 
 import cudf
 import cupy as cp
-from cudf.utils.dtypes import is_list_dtype
+#from cudf.utils.dtypes import is_list_dtype
```
Comment (benfred) [inline nvtabular/loader/backend.py:24]:
```suggestion
```

File: nvtabular/loader/backend.py [hunk 4]
```diff
@@ -104,37 +104,42 @@ def batch(self, itr):
                 yield current
                 current = []
 
+    def chunk_logic(self, itr):
```
Comment (benfred) [inline nvtabular/loader/backend.py:107]: overindented here?

Comment (jperez999) [reply nvtabular/loader/backend.py:106]: @benfred I want this in the ChunkQueue class... should it not be in the class itself? I have references to self inside the method.

File: nvtabular/loader/backend.py [hunk 5]
```diff
@@ -491,7 +495,8 @@ def _create_tensors(self, gdf):
         column_groups = (self.cat_names, self.cont_names, self.label_names)
         dtypes = (self._LONG_DTYPE, self._FLOAT32_DTYPE, self._FLOAT32_DTYPE)
         tensors = []
-        offsets = cudf.DataFrame()
+        # offsets = cudf.DataFrame()
```
Comment (benfred) [inline nvtabular/loader/backend.py:498]:
```suggestion
```

File: nvtabular/loader/tensorflow.py [hunk 6]
```diff
@@ -297,26 +302,40 @@ def _LONG_DTYPE(self):
     def _FLOAT32_DTYPE(self):
         return tf.float32
 
+    def _to_dlpack(self, gdf):
```
Comment (benfred) [inline nvtabular/loader/tensorflow.py:305]: This method doesn't always return a dlpack object - and I'm wondering if we should rename. Maybe something like ```def _pack(self, df)``` for this method and ```def _unpack(self, df)``` for the _from_dlpack method?

Also I'd like to stop using 'gdf' for cases where the dataframe can be either a pandas or cudf dataframe. Can we go with 'df' here instead?

File: nvtabular/dispatch.py [hunk 7]
```diff
@@ -167,3 +167,20 @@ def _to_arrow(x):
...
+
+
+def _concat(objs, **kwargs):
+    if isinstance(objs[0], (cudf.DataFrame, cudf.Series)):
+        return cudf.core.reshape.concat(objs, **kwargs)
+    else:
+        return pd.concat(objs, **kwargs)
+
+
+def _make_df(x=None, device=None):
```
Comment (benfred) [inline nvtabular/dispatch.py:179]: elsewhere in this file we're using ```_like_df``` to indicate what type of dataframe it should be based off (see like _arange etc). Can we go

```suggestion
def _make_df(_like_df=None, device=None):
```

to be consistent?

File: nvtabular/loader/torch.py [hunk 8]
```diff
@@ -107,11 +107,23 @@ def __iter__(self):
         return DataLoader.__iter__(self)
 
     def _get_device_ctx(self, dev):
+        if dev == "cpu":
+            return torch.device("cpu")
         return torch.cuda.device("cuda:{}".format(dev))
 
+    def _to_dlpack(self, gdf):
```
Comment (benfred) [inline nvtabular/loader/torch.py:114]: Can we also rename this too? (_pack/_unpack)

**Review by benfred — COMMENTED** [review]
Looks good!

**Review by benfred — APPROVED** [review]

**Pull request merged**